contracts/HubPool.sol

```diff
@@ -278,24 +289,85 @@ contract HubPool is Testable, Lockable, MultiCaller, Ownable {
-            "Passed liveness"
+    function disputeRelayerRefund() public {
+        require(getCurrentTime() <= refundRequest.requestExpirationTimestamp, "Passed liveness");
+
+        // Request price from OO and dispute it.
+        uint256 totalBond = _getBondTokenFinalFee() + bondAmount;
+        bondToken.safeTransferFrom(msg.sender, address(this), totalBond);
+        // This contract needs to approve totalBond*2 against the OO contract. (for the price request and dispute).
+        bondToken.safeApprove(address(_getOptimisticOracle()), totalBond * 2);
+        _getOptimisticOracle().requestAndProposePriceFor(
+            identifier,
+            uint32(getCurrentTime()),
+            _getRefundProposalAncillaryData(),
+            bondToken,
+            // Set reward to 0, since we'll settle proposer reward payouts directly from this contract after a relay
+            // proposal has passed the challenge period.
+            0,
+            // Set the Optimistic oracle proposer bond for the price request.
+            bondAmount,
+            // Set the Optimistic oracle liveness for the price request.
+            refundProposalLiveness,
+            refundRequest.proposer,
+            // Canonical value representing "True"; i.e. the proposed relay is valid.
+            int256(1e18)
         );
 
-        // Delete the last element in the relayerRefundRequests array. This acts to throw out the request.
-        emit RelayerRefundDisputed(relayerRefundRequestId, msg.sender);
+        // Dispute the request that we just sent.
+        SkinnyOptimisticOracleInterface.Request memory ooPriceRequest = SkinnyOptimisticOracleInterface.Request({
+            proposer: refundRequest.proposer,
+            disputer: address(0),
+            currency: bondToken,
+            settled: false,
+            proposedPrice: int256(1e18),
+            resolvedPrice: 0,
+            expirationTime: getCurrentTime() + refundProposalLiveness,
+            reward: 0,
+            finalFee: _getBondTokenFinalFee(),
+            bond: bondAmount,
+            customLiveness: refundProposalLiveness
+        });
+
+        _getOptimisticOracle().disputePriceFor(
+            identifier,
+            uint32(getCurrentTime()),
+            _getRefundProposalAncillaryData(),
+            ooPriceRequest,
+            msg.sender,
+            address(this)
+        );
 
-        relayerRefundRequests.pop();
+        emit RelayerRefundDisputed(msg.sender, ooPriceRequest, refundRequest);
 
-        // TODO: pull bonds. request price from OO.
+        // Finally, delete the state pertaining to the active refundRequest.
+        delete refundRequest;
     }
 
-    function getNumberOfRelayRefunds() public view returns (uint256) {
-        if (relayerRefundRequests.length == 0) return 0;
-        return relayerRefundRequests.length - 1;
+    function _getRefundProposalAncillaryData() public view returns (bytes memory ancillaryData) {
```

Comment:
this was my initial pass at constructing the ancillary data. curious what you guys think & if we need more/less in this.

Reply:
these seem reasonable. I'm thinking about ways we can reduce the amount of calls to `appendKeyValue...` because these functions are expensive I believe

Reply:
Do we call this in the happy path? Or only on disputes? If only on disputes, we might not need to worry about gas quite as much (although it's still worth thinking about!!).

Reply:
it only happens in the dispute case so the gas does not matter too much I don't think.